geomstats/geometry/complex_poincare_disk.py

```diff
@@ -0,0 +1,360 @@
+        inner_prod_mat *= self.scale**2
+        return inner_prod_mat
+
+    def inner_product(self, tangent_vec_a, tangent_vec_b, base_point):
+        """Compute the complex Poincaré disk inner-product.
+
+        Compute the inner-product of tangent_vec_a and tangent_vec_b
+        at point base_point using the Siegel Riemannian metric:
+
+        Parameters
+        ----------
+        tangent_vec_a : array-like, shape=[..., 1]
+            Tangent vector at base point.
+        tangent_vec_b : array-like, shape=[..., 1]
+            Tangent vector at base point.
+        base_point : array-like, shape=[..., 1]
+            Base point.
+
+        Returns
+        -------
+        inner_product : array-like, shape=[...,]
+            Inner-product.
+        """
+        inner_product_matrix = self.inner_product_matrix(base_point=base_point)
+        inner_prod = gs.conj(tangent_vec_a) * inner_product_matrix * tangent_vec_b
+        inner_prod = gs.reshape(inner_prod, (-1,))
+        return inner_prod
+
+    def squared_norm(self, vector, base_point):
+        """Compute the square of the norm of a vector.
+
+        Squared norm of a vector associated to the inner product
+        at the tangent space at a base point.
+
+        Parameters
+        ----------
+        vector : array-like, shape=[..., 1]
+            Vector.
+        base_point : array-like, shape=[..., 1]
+            Base point.
+            Optional, default: None.
+
+        Returns
+        -------
+        sq_norm : array-like, shape=[...,]
+            Squared norm.
+        """
+        sq_norm = self.inner_product(vector, vector, base_point=base_point)
+        sq_norm = gs.real(sq_norm)
+        return sq_norm
+
+    @staticmethod
+    def exp(tangent_vec, base_point):
+        """Compute the complex Poincaré disk exponential map.
+
+        Parameters
+        ----------
+        tangent_vec : array-like, shape=[..., 1]
+            Tangent vector at base point.
+        base_point : array-like, shape=[..., 1]
+            Base point.
+
+        Returns
+        -------
+        exp : array-like, shape=[..., 1]
+            Riemannian exponential.
+        """
+        ndim = max(len(gs.shape(tangent_vec)), len(gs.shape(base_point)))
+        tangent_vec = gs.to_ndarray(tangent_vec, to_ndim=2)
+        base_point = gs.to_ndarray(base_point, to_ndim=2)
+        if base_point.shape[0] < tangent_vec.shape[0]:
+            base_point = (
+                gs.cast(gs.ones(tangent_vec.shape), dtype=base_point.dtype) @ base_point
+            )
+        elif base_point.shape[0] > tangent_vec.shape[0]:
+            tangent_vec = (
+                gs.cast(gs.ones(base_point.shape), dtype=tangent_vec.dtype)
+                @ tangent_vec
+            )
+        theta = gs.cast(gs.angle(tangent_vec), dtype=gs.get_default_cdtype())
+        s = 2 * gs.abs(tangent_vec) / (1 - gs.abs(base_point) ** 2)
+        num = gs.array(base_point)
```

Comment:
`gs.copy`?

Reply:
If I use `num = gs.copy(base_point)`, then TensorFlow gives 7 failing tests:


`=================================================================================================== short test summary info ===================================================================================================
FAILED tests/tests_geomstats/test_complex_poincare_disk.py::TestComplexPoincareDiskMetric::test_exp[1.0-tangent_vec0-base_point0-expected0] - RuntimeError: `variable += value` with `tf.Variable`s is not supported. Use `v...
FAILED tests/tests_geomstats/test_complex_poincare_disk.py::TestComplexPoincareDiskMetric::test_exp[10.0-tangent_vec1-base_point1-expected1] - RuntimeError: `variable += value` with `tf.Variable`s is not supported. Use `...
FAILED tests/tests_geomstats/test_complex_poincare_disk.py::TestComplexPoincareDiskMetric::test_exp[1.0-tangent_vec2-base_point2-expected2] - RuntimeError: `variable += value` with `tf.Variable`s is not supported. Use `v...
FAILED tests/tests_geomstats/test_complex_poincare_disk.py::TestComplexPoincareDiskMetric::test_exp_after_log[connection_args0-point0-base_point0-1e-05-1e-06] - RuntimeError: `variable += value` with `tf.Variable`s is no...
FAILED tests/tests_geomstats/test_complex_poincare_disk.py::TestComplexPoincareDiskMetric::test_exp_after_log[connection_args1-point1-base_point1-1e-05-1e-06] - RuntimeError: `variable += value` with `tf.Variable`s is no...
FAILED tests/tests_geomstats/test_complex_poincare_disk.py::TestComplexPoincareDiskMetric::test_log_after_exp[connection_args0-tangent_vec0-base_point0-1e-05-1e-06] - RuntimeError: `variable += value` with `tf.Variable`s...
FAILED tests/tests_geomstats/test_complex_poincare_disk.py::TestComplexPoincareDiskMetric::test_log_after_exp[connection_args1-tangent_vec1-base_point1-1e-05-1e-06] - RuntimeError: `variable += value` with `tf.Variable`s...
====================================================================================`


Here is the details of the last failing test:


`___________________________________________________________ TestComplexPoincareDiskMetric.test_log_after_exp[connection_args1-tangent_vec1-base_point1-1e-05-1e-06] ___________________________________________________________

self = <tests.tests_geomstats.test_complex_poincare_disk.TestComplexPoincareDiskMetric object at 0x7fcfd2362d30>, connection_args = (4,)
tangent_vec = <tf.Tensor: shape=(3, 1), dtype=complex128, numpy=
array([[-1.59327767+1.68671547j],
       [-0.23285997+0.90774852j],
       [-1.05644183+0.30277497j]])>
base_point = <tf.Tensor: shape=(1,), dtype=complex128, numpy=array([0.01180741+0.00911528j])>, rtol = 1e-05, atol = 1e-06

    def test_log_after_exp(self, connection_args, tangent_vec, base_point, rtol, atol):
        """Check that connection exponential and logarithm are inverse.
    
        This is calling connection exponential first, then connection logarithm.
    
        Parameters
        ----------
        connection_args : tuple
            Arguments to pass to constructor of the connection.
        tangent_vec : array-like
            Tangent vector to the manifold at base_point.
        base_point : array-like
            Point on the manifold.
        rtol : float
            Relative tolerance to test this property.
        atol : float
            Absolute tolerance to test this property.
        """
        connection = self.Metric(*connection_args)
>       exp = connection.exp(tangent_vec=tangent_vec, base_point=gs.array(base_point))

tests/geometry_test_cases.py:823: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
geomstats/geometry/complex_poincare_disk.py:228: in exp
    num += exp_i_theta
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <tf.Variable 'Variable:0' shape=(3, 1) dtype=complex128, numpy=
array([[0.01180741+0.00911528j],
       [0.01180741+0.00911528j],
       [0.01180741+0.00911528j]])>
unused_other = <tf.Tensor: shape=(3, 1), dtype=complex128, numpy=
array([[-0.68668472+0.72695536j],
       [-0.24847941+0.96863718j],
       [-0.96129902+0.27550715j]])>

    def __iadd__(self, unused_other):
>     raise RuntimeError("`variable += value` with `tf.Variable`s is not "
                         "supported. Use `variable.assign_add(value)` to modify "
                         "the variable, or `out = variable + value` if you "
                         "need to get a new output Tensor.")
E     RuntimeError: `variable += value` with `tf.Variable`s is not supported. Use `variable.assign_add(value)` to modify the variable, or `out = variable + value` if you need to get a new output Tensor.

venv/lib/python3.8/site-packages/tensorflow/python/ops/resource_variable_ops.py:1472: RuntimeError`

Reply:
Done.